### PR TITLE
Make defense more aggressive in stealing balls from enemy robots

### PR DESCRIPTION
**Pull request by williamckha**
### Description

Makes defense play more aggressive by adding shadowers that can steal balls from enemies. 

The defense play will switch to `AggressiveDefenseState` when the possession tracker sees that the enemy team has had possession for a while (possession is stagnant). The time threshold which determines stagnancy is a configurable param that we can adjust (shorter time -> more aggressive). In `AggressiveDefenseState`, the pass defender targeting the primary enemy is replaced with a shadower so that we can attempt to steal the ball.

### Review Checklist

**_It is the reviewers responsibility to also make sure every item here has been covered_**

- [ ] **Function & Class comments**: All function definitions (usually in the `.h` file) should have a javadoc style comment at the start of them. For examples, see the functions defined in `thunderbots/software/geom`. Similarly, all classes should have an associated Javadoc comment explaining the purpose of the class.
- [ ] **Remove all commented out code**
- [ ] **Remove extra print statements**: for example, those just used for testing
- [ ] **Resolve all TODO's**: All `TODO` (or similar) statements should either be completed or associated with a github issue

### CODE REVIEW

**Review by sauravbanna — DISMISSED** [review]
LGTM

File: src/software/ai/hl/stp/play/defense/defense_play_fsm.cpp [hunk 1]
```diff
@@ -144,3 +180,29 @@ void DefensePlayFSM::setUpPassDefenders(unsigned int num_pass_defenders)
...
+{
+    if (num_shadowers == shadowers.size())
+    {
+        return;
+    }
+
+    shadowers = std::vector<std::shared_ptr<ShadowEnemyTactic>>(num_shadowers);
+    std::generate(shadowers.begin(), shadowers.end(),
+                  [this]() { return std::make_shared<ShadowEnemyTactic>(); });
+}
```
Comment (itsarune) [inline src/software/ai/hl/stp/play/defense/defense_play_fsm.cpp:194]: if we're close enough to a robot, what happens if we try assigning a dribble tactic to actually try to steal the ball from the robot?

File: src/software/sensor_fusion/possession/possession_tracker.cpp [hunk 2]
```diff
@@ -36,7 +37,14 @@ TeamPossession PossessionTracker::getTeamWithPossession(const Team &friendly_tea
     else if ((time_near_friendly < time_near_threshold) &&
              (time_near_enemy > time_near_threshold))
     {
-        possession = TeamPossession::ENEMY_TEAM;
+        if (time_near_enemy < time_stagnant_threshold)
```
Comment (itsarune) [inline src/software/sensor_fusion/possession/possession_tracker.cpp]: I feel like stagnant position should also be tied to ball position. It'd be more appropriate to transition to aggressive, ball-stealing defence if the ball hasn't moved considerably between subsequent ticks

File: src/software/ai/hl/stp/play/defense/defense_play_fsm.cpp (hunk 1 above)
Comment (williamckha) [reply src/software/ai/hl/stp/play/defense/defense_play_fsm.cpp:194]: I believe the shadow enemy tactic tries to dribble into the ball and steal it after the shadower gets into position

File: src/software/sensor_fusion/possession/possession_tracker.cpp (hunk 2 above)
Comment (williamckha) [reply src/software/sensor_fusion/possession/possession_tracker.cpp]: Added a check for ball stagnancy

**Comment by williamckha** [comment]
Closing this, we will look into better ways of managing our defense as part of gameplay overhaul

**Comment by nimazareian** [comment]
We should consider reintegrating this back into our current gameplay as defense is not a focus of the new gameplay architecture.

@itsarune @williamckha What do you think?

**Comment by williamckha** [comment]
I figure it won't be too difficult to convert our defense over to the new gameplay. The DefensePlay can have a designated "primary defender" tactic (equivalent of AttackerTactic in offensive plays) and we can have defensive skills that the primary defender can execute (e.g. block lane, shadow, steal, etc.). The advantage would be that our offense can reuse some skills (e.g. steal) and all our main gameplay would use the same architecture.

Once we've merged the new gameplay changes into master, I can open some tickets for converting our defense. The work should be easily parallelizable (writing and testing skills) so multiple team members can work on the defense conversion at the same time

**Pull request closed without merging**